File: United_Community/Mradul/Daily-Leetcode/day_2_TrappingRainWater.java
//https://leetcode.com/problems/trapping-rain-water/
class Solution {
    public int trap(int[] h) {
        int n = h.length, water = 0;
        if (n < 2) {
            return 0;
        }
        int[] LeftMax = new int[n];
        int[] RightMax = new int[n];

        LeftMax[0] = h[0];
        RightMax[n-1] = h[n-1];

        for (int i = 1; i < n; i++) {
            LeftMax[i] = Math.max(LeftMax[i-1], h[i]);
        }

        for (int i = n-2; i >= 0; i--) {
            RightMax[i] = Math.max(RightMax[i+1], h[i]);
        }

        for (int i = 1; i < n-1; i++) {
            water += (Math.min(RightMax[i], LeftMax[i]) - h[i]);
        }

        return water;
    }
}

File: United_Community/Amit/Daily-Leetcode/day_2_trapping_rain_water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        if (n <= 2) {
            return 0;
        }

        int trap_water = 0;

        vector<int> l_max(n, 0);
        vector<int> r_max(n, 0);

        l_max[0] = height[0];
        for (int i = 1; i < n; ++i) {
            l_max[i] = max(l_max[i - 1], height[i]);
        }

        r_max[n - 1] = height[n - 1];
        for (int i = n - 2; i >= 0; --i) {
            r_max[i] = max(r_max[i + 1], height[i]);
        }

        for (int i = 1; i < n - 1; ++i) {
            trap_water += max(0, min(l_max[i], r_max[i]) - height[i]);
        }

        return trap_water;
    }
};
